### PR TITLE
Issue #11259: Add record support to RedundantModifier

### LINKED ISSUE #11259: RedundantModifier should include RECORD_DEF in acceptable tokens
Issue property reported by @Vyom-Yadav

I have read check documentation: http://checkstyle.sourceforge.net/config_modifier.html#RedundantModifier
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
/var/tmp $ javac Test.java

/var/tmp $ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="RedundantModifier"/>
    </module>
</module>

/var/tmp $ cat Test.java
final record foo() {

    static final record hoo() {
    }
}


/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-10.0-all.jar -c config.xml YOUR_FILE.java
Starting audit...
Audit done.
```

---

**Describe what you expect in detail.**
```
/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-10.0-all.jar -c config.xml YOUR_FILE.java
Starting audit...
[ERROR] /var/tmp/Test.java:1:1: Redundant 'final' modifier. [RedundantModifier]
[ERROR] /var/tmp/Test.java:3:5: Redundant 'static' modifier. [RedundantModifier]
[ERROR] /var/tmp/Test.java:3:12: Redundant 'final' modifier. [RedundantModifier]
Audit done.
Checkstyle ends with 3 errors.
```

Related info https://docs.oracle.com/en/java/javase/17/language/records.html

### PR DESCRIPTION
**Pull request by Vyom-Yadav**
Resolves #11259 
Diff Regression config: https://gist.githubusercontent.com/Vyom-Yadav/98dceb63a79f4833e85fff9b2e1464a6/raw/2984d7a7e418099c61e407c3dd56748ba5cada87/my_checks.xml

Diff Regression patch config: https://gist.githubusercontent.com/Vyom-Yadav/98dceb63a79f4833e85fff9b2e1464a6/raw/2984d7a7e418099c61e407c3dd56748ba5cada87/my_checks.xml

### CODE REVIEW

**Comment by Vyom-Yadav** [comment]
Github, generate report

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java
Comment (strkkk) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java:324]: I think it can be inlined, no need to extract separate method

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java
Comment (strkkk) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java:82]: please add some link to jdk spec (or something similar) to make it clearer for users

**Review by strkkk — CHANGES_REQUESTED** [review]

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java
Comment (Vyom-Yadav) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java:264]: Ok, refactor both `if-else` chain to switch? (main + `if-else` nested in `else`)

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java
Comment (Vyom-Yadav) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java
Comment (Vyom-Yadav) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java]: Done.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java]: no, lets leave it as it is, to avoid extending diff with unrelated changes

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java:265]: yes. `switch` is designed exactly for cases like this - to avoid long if-else chains. Plus, here "nested" if else is nothing but a continuation of outer if-else chain.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java
Comment (strkkk) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java:83]: Please make it like `See <link> for more info`

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java
Comment (Vyom-Yadav) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java:265]: yes for refactoring both chains to switch **and** for extraction?

Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java:265]: yes, to both.

Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java:265]: it also may be a second commit with refactoring

Comment (Vyom-Yadav) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java]: I changed it, kindly take a look.

**Review by strkkk — APPROVED** [review]

**Comment by nmancus1** [comment]
Github, generate site

**Comment by nmancus1** [comment]
Link to check documentation: https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/c8ea5d8_2022132949/config_modifier.html#RedundantModifier

**Review by nmancus1 — CHANGES_REQUESTED** [review]
Thoughts:

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java
Comment (Vyom-Yadav) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java:281]: I converted `if-else` to `switch` without changing the current logic of check, the current logic contains a lot of false negatives, I have opened #11268, will send a fix for it, once that is merged, I will complete this PR.

**Comment by nmancus1** [comment]
Blocked by https://github.com/checkstyle/checkstyle/issues/11268, we can do refactoring of `visitToken` there, then continue work in this PR.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java:281]: > will send a fix for it, once that is merged, I will complete this PR

Added blocked label, and approved https://github.com/checkstyle/checkstyle/issues/11268

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java:242]: there can be issues with coverage, since this impossible case will be impossible to test

Comment (Vyom-Yadav) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java:242]: We can create a test for testing non acceptable token-
```java
@Test                                                                                  
public void testWrongTokenType() {                                                     
    final RedundantModifierCheck obj = new RedundantModifierCheck();                   
    final DetailAstImpl ast = new DetailAstImpl();                                     
    ast.initialize(new CommonToken(TokenTypes.LITERAL_NULL, "null"));                  
    try {                                                                              
        obj.visitToken(ast);                                                           
        assertWithMessage("IllegalStateException is expected").fail();                 
    }                                                                                  
    catch (IllegalStateException ex) {                                                 
        assertWithMessage("Invalid exception message")                                  
                .that(ex.getMessage())                                                  
                .isEqualTo("Unexpected value: " + ast.getType());                      
    }                                                                                  
}                                                                                      
```

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java
Comment (Vyom-Yadav) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java:441]: Can change `TokenUtil.forEachChild` to accept multiple types but that didn't feel right as most of the time you would want to do a different operation to different types of tokens.

**Comment by Vyom-Yadav** [comment]
Github, generate report

**Comment by Vyom-Yadav** [comment]
Github, generate site

**Comment by Vyom-Yadav** [comment]
https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/93b922f_2022041314/config_modifier.html#RedundantModifier

**Review by nick-mancuso — APPROVED** [review]

**Review by pbludov — APPROVED** [review]

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java
Comment (nick-mancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java]: Resolved

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java
Comment (nick-mancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java]: Resolved

**Comment by romani** [comment]
@Vyom-Yadav, please put a comment in issue that is a form how issue should be reported.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java:82]: >Nested records are implicitly static.

I do not see reflection of this in code.
Please make input for not nested record.
Please extend violation comments in Input to have message part that define what modifier is redundant.

**Review by romani — CHANGES_REQUESTED** [review]
Items:

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java
Comment (Vyom-Yadav) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java:82]: > > Nested records are implicitly static.
> 
> I do not see reflection of this in code. Please make input for not nested record. 

There is a top-level record in the example, top-level records can't have the keyword static (compile-time error), examples for nested records are there.

> Please extend violation comments in Input to have message part that define what modifier is redundant.

Added where there was only a single violation.

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/redundantmodifier/InputRedundantModifierRecords.java
Comment (romani) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/redundantmodifier/InputRedundantModifierRecords.java:38]: Add one more top level record, that has no violations as example what is good form of  code.
Our inputs should be combination of valid and invalid input code.

**Review by romani — CHANGES_REQUESTED** [review]
Items:

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/redundantmodifier/InputRedundantModifierRecords.java
Comment (Vyom-Yadav) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/redundantmodifier/InputRedundantModifierRecords.java:38]: done.

**Review by romani — APPROVED** [review]
Ok to merge

**Review by rnveach — APPROVED** [review]

**Comment by nick-mancuso** [comment]
Failure in wercker: `get of docker network name failed: API error (404): {"message":"could not find an available, non-overlapping IPv4 address pool among the defaults to assign to the network"}`

From https://app.wercker.com/checkstyle/checkstyle/runs/build/6224e2c2c7d24600072a6659.

Multiple retries have failed with the same.

**Comment by rnveach** [comment]
We also have spelling errors being reported.

**Comment by Vyom-Yadav** [comment]
@rnveach Only wercker is failing now.

**Comment by romani** [comment]
Github, rebase

**Pull request merged**
